design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,205 @@
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+### Syntax
+
+Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
+actions that simply declare metadata on top of existing data. This solves two issues:
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+In the future, there can exist shortcuts (similar to `C*AS`) to both create connectors and declare 
+corresponding streams/tables.
+
+There are a few more minor points of consideration:
+- All "DDL" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- To make this work for all vanilla connectors, we will need Connect to implement an API that shows
+  all topics created by a connector.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
+
+--------------
+SOURCE CREATED
+--------------
+```
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
+
+name: my-postgres-jdbc
+connector: io.confluent.connect.jdbc.SourceConnector
+topic-prefix: test-postgres-jdbc
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
```

Comment:
Where is the schema coming from here? Does the Connect API have enough detail to provide this for non-Avro formats?

Reply:
No it does not - this functionality would only work when it auto-registers with SchemaRegistry. For non-Avro formats, Connect would need to be extended to support exposing a schema API.